File: 6. Uso de Estruturas de Dados Pre-Definidas/Supermercado em C++/sistema.cpp
#include "sistema.hpp"
#include <iostream>
#include <iomanip>
#include <algorithm>
using namespace std;

Sistema::Sistema() {
    compraAtual = nullptr;
}

Sistema::~Sistema() {
    for (auto produto : produtos) {
        delete produto;
    }
    for (auto cliente : clientes) {
        delete cliente;
    }
    for (auto compra : compras) {
        delete compra;
    }
    if (compraAtual != nullptr) {
        delete compraAtual;
    }
}

void Sistema::registrarProduto(string nome, double preco) {
    Produto* produto = new Produto(nome, preco);
    produtos.push_back(produto);
}

void Sistema::registrarCliente(string nome, string cpf) {
    Cliente* cliente = new Cliente(nome, cpf);
    clientes.push_back(cliente);
}

void Sistema::adquirirProduto(string nomeProduto, int quantidade) {
    for (auto produto : produtos) {
        if (produto->getNome() == nomeProduto) {
            produto->adicionarEstoque(quantidade);
            return;
        }
    }
}

void Sistema::iniciarCompra(string cpf) {
    for (auto cliente : clientes) {
        if (cliente->getCpf() == cpf) {
            compraAtual = new Compra(cliente);
            return;
        }
    }
}

void Sistema::addItem(string nomeProduto, int quantidade) {
    if (compraAtual != nullptr) {
        for (auto produto : produtos) {
            if (produto->getNome() == nomeProduto) {
                compraAtual->adicionarProduto(produto, quantidade);
                return;
            }
        }
    }
}

void Sistema::finalizarCompra() {
    if (compraAtual != nullptr) {
        double total = compraAtual->calcularTotal();
        cout << fixed << setprecision(2) << total << endl;
        compraAtual->finalizarCompra();
        compras.push_back(compraAtual);
        compraAtual = nullptr;
    }
}

void Sistema::revenue() {
    double total = 0;
    for (auto compra : compras) {
        total += compra->getValorTotal();
    }
    cout << fixed << setprecision(2) << total << endl;
}

void Sistema::maisVendido() {
    vector<pair<Produto*, int>> vendas;

    for (auto compra : compras) {
        vector<pair<Produto*, int>> produtosCompra = compra->getProdutos();
        for (auto& par : produtosCompra) {
            bool encontrado = false;
            for (auto& venda : vendas) {
                if (venda.first == par.first) {
                    venda.second += par.second;
                    encontrado = true;
                    break;
                }
            }
            if (!encontrado) {
                vendas.push_back(make_pair(par.first, par.second));
            }
        }
    }

    Produto* maisVendido = nullptr;
    int maxVendas = 0;

    for (auto& par : vendas) {
        if (par.second > maxVendas) {
            maxVendas = par.second;
            maisVendido = par.first;
        }
    }

    if (maisVendido != nullptr) {
        cout << maisVendido->getNome() << endl;
    }
}

void Sistema::melhorCliente() {
    vector<pair<Cliente*, int>> numCompras;

    for (auto compra : compras) {
        bool encontrado = false;
        for (auto& par : numCompras) {
            if (par.first == compra->getCliente()) {
                par.second++;
                encontrado = true;
                break;
            }
        }
        if (!encontrado) {
            numCompras.push_back(make_pair(compra->getCliente(), 1));
        }
    }

    Cliente* melhorCliente = nullptr;
    int maxCompras = 0;

    for (auto& par : numCompras) {
        if (par.second > maxCompras) {
            maxCompras = par.second;
            melhorCliente = par.first;
        }
    }

    if (melhorCliente != nullptr) {
        cout << melhorCliente->getNome() << endl;
    }
}

void Sistema::emFalta() {
    for (auto produto : produtos) {
        if (produto->getQuantidade() < 5) {
            cout << produto->getNome() << endl;
        }
    }
}

File: 6. Uso de Estruturas de Dados Pre-Definidas/Supermercado em Java/Sistema.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

public class Sistema {
    private HashMap<String, Produto> produtos;
    private HashMap<String, Cliente> clientes;
    private ArrayList<Compras> compras;
    private Compras compraAtual;

    public Sistema() {
        this.produtos = new HashMap<>();
        this.clientes = new HashMap<>();
        this.compras = new ArrayList<>();
        this.compraAtual = null;
    }

    public void registrarProduto(String nome, double preco) {
        Produto produto = new Produto(nome, preco);
        produtos.put(nome, produto);
    }

    public void registrarCliente(String nome, String cpf) {
        Cliente cliente = new Cliente(nome, cpf);
        clientes.put(cpf, cliente);
    }

    public void adquirirProduto(String nomeProduto, int quantidade) {
        Produto produto = produtos.get(nomeProduto);
        if (produto != null) {
            produto.adicionarEstoque(quantidade);
        }
    }

    public void iniciarCompra(String cpf) {
        Cliente cliente = clientes.get(cpf);
        if (cliente != null) {
            compraAtual = new Compras(cliente);
        }
    }

    public void addItem(String nomeProduto, int quantidade) {
        Produto produto = produtos.get(nomeProduto);
        if (produto != null && compraAtual != null) {
            compraAtual.adicionarProduto(produto, quantidade);
        }
    }

    public void finalizarCompra() {
        if (compraAtual != null) {
            double total = compraAtual.calcularTotal();
            System.out.printf(Locale.US,"%.2f%n", total);
            compraAtual.finalizarCompra();
            compras.add(compraAtual);
            compraAtual = null;
        }
    }

    public void revenue() {
        double total = 0;
        for (Compras compra : compras) {
            total += compra.getValorTotal();
        }
        System.out.printf(Locale.US,"%.2f%n", total);
    }

    public void maisVendido() {
        HashMap<Produto, Integer> vendas = new HashMap<>();
        
        for (Compras compra : compras) {
            for (Map.Entry<Produto, Integer> entry : compra.getProdutos().entrySet()) {
                Produto produto = entry.getKey();
                int quantidade = entry.getValue();
                vendas.put(produto, vendas.getOrDefault(produto, 0) + quantidade);
            }
        }

        Produto maisVendido = null;
        int maxVendas = 0;
        
        for (Map.Entry<Produto, Integer> entry : vendas.entrySet()) {
            if (entry.getValue() > maxVendas) {
                maxVendas = entry.getValue();
                maisVendido = entry.getKey();
            }
        }

        if (maisVendido != null) {
            System.out.println(maisVendido.getNome());
        }
    }

    public void melhorCliente() {
        HashMap<Cliente, Integer> numCompras = new HashMap<>();
        
        for (Compras compra : compras) {
            Cliente cliente = compra.getCliente();
            numCompras.put(cliente, numCompras.getOrDefault(cliente, 0) + 1);
        }

        Cliente melhorCliente = null;
        int maxCompras = 0;
        
        for (Map.Entry<Cliente, Integer> entry : numCompras.entrySet()) {
            if (entry.getValue() > maxCompras) {
                maxCompras = entry.getValue();
                melhorCliente = entry.getKey();
            }
        }

        if (melhorCliente != null) {
            System.out.println(melhorCliente.getNome());
        }
    }

    public void emFalta() {
        for (Produto produto : produtos.values()) {
            if (produto.getQtd() < 5) {
                System.out.println(produto.getNome());
            }
        }
    }
}